### PR TITLE
Service account/roles/binding functions

### PR DESCRIPTION
**Pull request by dprince**

### CODE REVIEW

File: modules/common/condition/conditions.go [hunk 1]
```diff
@@ -247,3 +257,25 @@ const (
...
+    RoleReadyErrorMessage = "Role error occurred %s"
+
+    // RoleCreatingMessage
+    RoleCreatingMessage = "Role creation in progress"
+
+    // RoleBindingReadyErrorMessage
+    RoleBindingReadyErrorMessage = "RoleBinding error occurred %s"
+
+    // RoleBindingCreatingMessage
+    RoleBindingCreatingMessage = "RoleBinding creation in progress"
```
Comment (stuggi) [inline modules/common/condition/conditions.go:280]: `make gofmt`, different indentation ?

File: modules/common/role/role.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+package role
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/util"
+    rbacv1 "k8s.io/api/rbac/v1"
```
Comment (stuggi) [inline modules/common/role/role.go:26]: `make gofmt`, different indentation ?

File: modules/common/serviceaccount/serviceaccount.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// NewServiceAccount returns an initialized ServiceAccount
+func NewServiceAccount(
+	serviceAccount *corev1.ServiceAccount,
+	labels map[string]string,
```
Comment (stuggi) [inline modules/common/serviceaccount/serviceaccount.go:37]: not used?

File: modules/common/serviceaccount/serviceaccount.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.serviceAccount.Name,
+			Namespace: s.serviceAccount.Namespace,
+		},
+	}
+
+	op, err := controllerutil.CreateOrPatch(ctx, h.GetClient(), sa, func() error {
+		sa.Labels = util.MergeStringMaps(sa.Labels, s.serviceAccount.Labels)
+		sa.Annotations = s.serviceAccount.Annotations
```
Comment (stuggi) [inline modules/common/serviceaccount/serviceaccount.go:60]: merge annotations?

File: modules/common/serviceaccount/serviceaccount.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		}
+
+		return nil
+	})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			h.GetLogger().Info(fmt.Sprintf("ServiceAccount %s not found, reconcile in %s", sa.Name, s.timeout))
+			return ctrl.Result{RequeueAfter: s.timeout}, nil
+		}
+		return ctrl.Result{}, err
```
Comment (stuggi) [inline modules/common/serviceaccount/serviceaccount.go:74]: might want to wrap the error to make it easier when debugging to know where it fails?

File: modules/common/rolebinding/types.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"time"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+// RoleBinding -
+type RoleBinding struct {
+	roleBinding *rbacv1.RoleBinding
+	timeout        time.Duration
```
Comment (stuggi) [inline modules/common/rolebinding/types.go:28]: fmt?

File: modules/common/serviceaccount/serviceaccount.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	labels map[string]string,
+	timeout time.Duration,
+) *ServiceAccount {
+	return &ServiceAccount{
+		serviceAccount: serviceAccount,
+		timeout:        timeout,
+	}
+}
+
+// CreateOrPatch - creates or patches a route, reconciles after Xs if object won't exist.
```
Comment (stuggi) [inline modules/common/serviceaccount/serviceaccount.go:46]: `route` -> `service account`

File: modules/common/serviceaccount/serviceaccount.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+// Delete - delete a service.
+func (s *ServiceAccount) Delete(
+	ctx context.Context,
+	h *helper.Helper,
+) error {
+
+	err := h.GetClient().Delete(ctx, s.serviceAccount)
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		err = fmt.Errorf("Error deleting serviceAccount %s: %v", s.serviceAccount.Name, err)
```
Comment (stuggi) [inline modules/common/serviceaccount/serviceaccount.go:91]: `%v` -> `%w`

File: modules/common/role/role.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+// NewRole returns an initialized Role
+func NewRole(
+	role *rbacv1.Role,
+	labels map[string]string,
+	timeout time.Duration,
+) *Role {
+	return &Role{
+		role: role,
+		timeout:        timeout,
```
[the remark this answers is not in the document]
Comment (stuggi) [reply modules/common/role/role.go:42]: I have proposed https://github.com/openstack-k8s-operators/lib-common/pull/239 which then also runs fmt check

File: modules/common/role/role.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+	}
+}
+
+// CreateOrPatch - creates or patches a route, reconciles after Xs if object won't exist.
```
[the remark this answers is not in the document]
Comment (dprince) [reply modules/common/role/role.go:46]: fixed.

File: modules/common/serviceaccount/serviceaccount.go (hunk 8 above)
Comment (dprince) [reply modules/common/serviceaccount/serviceaccount.go:91]: fixed.

File: modules/common/serviceaccount/serviceaccount.go (hunk 7 above)
Comment (dprince) [reply modules/common/serviceaccount/serviceaccount.go:46]: fixed.

File: modules/common/role/role.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// NewRole returns an initialized Role
+func NewRole(
+	role *rbacv1.Role,
+	labels map[string]string,
```
[the remark this answers is not in the document]
Comment (dprince) [reply modules/common/role/role.go:37]: it looks like we have a lot of objects that don't actually use these labels yet. I can wire them in...

File: modules/common/role/role.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.role.Name,
+			Namespace: r.role.Namespace,
+		},
+	}
+
+	op, err := controllerutil.CreateOrPatch(ctx, h.GetClient(), role, func() error {
+		role.Labels = util.MergeStringMaps(role.Labels, r.role.Labels)
+		role.Annotations = r.role.Annotations
```
[the remark this answers is not in the document]
Comment (dprince) [reply modules/common/role/role.go]: done in the latest commit

File: modules/common/serviceaccount/serviceaccount.go (hunk 4 above)
Comment (dprince) [reply modules/common/serviceaccount/serviceaccount.go]: done in the latest commit

File: modules/common/serviceaccount/serviceaccount.go (hunk 5 above)
Comment (dprince) [reply modules/common/serviceaccount/serviceaccount.go]: done

**Review by abays — APPROVED** [review]
/lgtm

**Pull request merged**